### PR TITLE
Add a BuildableFunctionView allowing statically rendered function views.

### LINKED ISSUE #130: Allow rendering function views
It's useful to be able to render function views as well as class views, models and templates.

This would allow greater flexibility to do things such as solve issue #88 (the reason why I initially went down this road).

I've used a similar method to solve rendering wagtail sitemaps statically [here](https://gitlab.com/ndevox/nick.sarbicki.com/blob/master/website/views.py#L9).

I'm submitting a PR (as I write this) which does something similar but is a bit more flexible and allows for different types of Django view return types (strings or HttpResponses).

### PR DESCRIPTION
**Pull request by NDevox**
There are some cases where building off of models and templates isn't
workable. A good example of this is a sitemap.xml.

This commit brings in a class which can render most Django function
views assuming it returns either a String or a HttpResponse.

Fixes #130.

### CODE REVIEW

**Comment by palewire** [comment]
Can you give ma a use case for this? Why not just make a subclass of Django's sitemap framework?

**Comment by NDevox** [comment]
I should have been more specific sorry.

I've used this for generating sitemaps in wagtail. This is the sitemap framework but ultimately ends as a callable you can use. Can see how I've done something very similar to this here: https://gitlab.com/ndevox/nick.sarbicki.com/blob/master/website/views.py#L16

Main reason for contributing this in is because I'd assumed this would have a far wider use case, e.g. any function view including self made ones or those provided by a third party package in use.

**Comment by palewire** [comment]
I could be caught up on pedantics, but the name here doesn't feel right to me. Something about a class being called a function has me thrown. I wonder we're not addressing the core issue, which I'm still not totally clear on. 

I think I'd prefer to address this by taking on the sitemaps framework more directly. Would it be a large hassle to rework this PR to frame the additions around the terminology of contrib.sitemaps?

**Comment by NDevox** [comment]
I think we're getting caught up in the sitemaps bit.

This could apply to any function, not just a sitemap build. It just so happens the function I've specifically made this class for initially was for sitemaps within wagtail.

If we were to make this specific to sitemaps I'd call it a `BuildableSitemapView` but that would significantly restrict the scope of what it can do.

**Comment by palewire** [comment]
Got it. Sorry to be obtuse. Can you explain to me the wider use case?

**Comment by NDevox** [comment]
No worries! I don't think I've been explaining it right.

I think the easiest way is essentially a view which requires some arbitrary data - not necessarily model data - which is done using a function view.

e.g. If you wanted a stats view that you only needed to update once every day. You could have a function which collects the stats (e.g. page views, number of users, users added in the last 24 hours etc.) and then creates a static page with this. Combine this with a cron job and you'd have a very simple way of generating a stats page.

Alternatively if you have a view which is going to combine multiple models in some way - where the list/detail views already available aren't perfectly suited. Again this could easily be done with a function view.

Or if you have other sets of data in your db outside of django models (not very common but happens). Again there isn't really anything catered for this currently in django-bakery but a function view could do it pretty easily.

Basically anything that doesn't rely on models, but requires more than just a static template.

**Comment by palewire** [comment]
I see. In my view, I'd prefer for our fundamental base views in the app to be able to handle this task, rather than create a new parallel system. What about the existing TemplateView doesn't serve your needs?

**Comment by NDevox** [comment]
So the other use case I should have mentioned is that there are many packages around for Django which will provide a function view for some core functionality. This is where I get the wagtail sitemap from - it's a function view they provide.

With the BuildableTemplateView I'd be doing (for my wagtail sitemap):

```python3
from wagtail.contrib.sitemaps.views import sitemap


class SitemapTemplateView(BuildableTemplateView):

    build_path = 'sitemap.xml'
    template_path = 'sitemap.xml'
    def get(self, request):
        return sitemap(self.request).render().content
```

This feels dirtier to me. I wouldn't supply a real template or context in this case because the function I've been given by the package I'm using already handles all of that, that would be rebuilding pre-existing functionality - sometimes quite complex functionality. So I have to give bogus values to make sure we don't error and override the get method to force return what comes out of this function, I'm working around the current functionality instead of with it.

With the proposed changes it would be:

```python3
from wagtail.contrib.sitemaps.views import sitemap


class SitemapBuildableView(BuildableFunctionView):
    function = sitemap
    build_path = 'sitemap.xml'
```

In some cases I can see the `BuildableTemplateView` working really well - namely for class based views that have a `get`. If I wasn't having to use function views from other packages I'd be happy to create views which simply do all the logic in the `get` method instead of using function views. It just felt quite dirty doing it with functions I'm provided with.

That said I do think there are plenty of cases for using functions over classes - e.g. views that don't require any of the additional functionality and complexity a class based view brings with it. This is where the proposed changes would come in useful in my mind.

I've got to admit, in the wider scheme of things, this isn't a massive deal. But it is quite convenient to have and makes some things far easier to bake for a newbie.

No offence taken if you don't want to bring it in though! I just thought it was nice to have the option.

**Comment by palewire** [comment]
I think I understand the problem, and feel free to call me a jerk here, but I think I still come down on keeping the overall architecture of the project the way it is and avoiding introducing a new idiom for rendering views.

An approach that appeals to me is two-pronged. One: A straightforward subclass of the existing Django sitemap views that add features to make them building, much in the way we've extended the Feeds framework. Two: Some lower-level refactoring of the building class-based views that yields a new base object that has the simpler feature set you're after.

**Comment by NDevox** [comment]
Sorry I dropped off the radar for this one.

It's obviously been a while, but the second approach might make some sense - I'd have to have a look at where we would do this.

The first approach is too narrow in scope IMO, it would solve the issue of generating sitemaps, but would not solve the issue of generating function based views.
